### PR TITLE
Reduce Warnings - Address warnings in examples and redundant recompilations

### PR DESCRIPTION
**Pull request by eramongodb**
Verified by [this patch](https://spruce.mongodb.com/version/66d0c133687c8500072caab4/).

Addresses 583 Clang warnings (157 unique) and 198 (7 unique) MSVC warnings relating to those emitted by examples (as part of CXX-2827). If a given warning was present in examples, this PR addresses it across all code for consistency.

404 Clang warnings and 191 MSVC warnings are addressed simply by reducing redundant recompilation of the new `client_helpers` and `spec_test_common` object library components by various mongocxx test executables. None of these warnings were unique: they were just noisy repetition of the same warning. This refactor should reduce both the warning count and the overall compilation time for mongocxx test executables.

Warnings pertaining to `__cdecl` are deferred to CXX-3092.

Warnings related to `_CRT_SECURE_NO_WARNINGS` with MSVC (getenv and gmtime) are left as-is. Feedback on how to approach these warnings (disable? `#if` blocks? a platform-dependent `BSONCXX_PRIVATE_GETENV` macro?) is welcome.

### CODE REVIEW

**Review by joshbsiegel — APPROVED** [review]
LGTM

**Review by kevinAlbs — COMMENTED** [review]
The warning fixes are much appreciated.

File: examples/mongocxx/automatic_client_side_field_level_encryption.cpp [hunk 1]
```diff
@@ -92,15 +94,18 @@ bsoncxx::document::value doc_from_file(std::string path) {
...
 
+}  // namespace
+
 int main() {
     instance inst{};
 
     // This must be the same master key that was used to create
     // the encryption key; here, we use a random key as a placeholder.
     char key_storage[kKeyLength];
     std::generate_n(key_storage, kKeyLength, std::rand);
```
[the remark this answers is not in the document]
Comment (eramongodb) [reply examples/mongocxx/automatic_client_side_field_level_encryption.cpp]: Replaced the raw `std::rand` (we shouldn't be addressing stdlib functions anyways) with a lambda expression performing the suggested cast instead.

File: examples/bsoncxx/view_and_value.cpp [hunk 2]
```diff
@@ -29,6 +29,10 @@
 #include <bsoncxx/types.hpp>
 #include <bsoncxx/types/bson_value/view.hpp>
 
+#include <bsoncxx/config/prelude.hpp>
```
[the remark this answers is not in the document]
Comment (eramongodb) [reply examples/bsoncxx/view_and_value.cpp]: Good suggestion. Done as suggested.

**Review by kevinAlbs — APPROVED** [review]

**Pull request merged**